miniapps/dpg/util/complexweakform.hpp

```diff
@@ -0,0 +1,276 @@
+   Array<Array<LinearFormIntegrator * > * > lfis_i;
+
+   BlockMatrix * P = nullptr; // Block Prolongation
+   BlockMatrix * R = nullptr; // Block Restriction
+
+   mfem::Operator::DiagonalPolicy diag_policy;
+
+   void Init();
+   void ReleaseInitMemory();
+
+   // Allocate appropriate SparseMatrix and assign it to mat
+   void AllocMat();
+
+   void ConformingAssemble();
+
+   void ComputeOffsets();
+
+   virtual void BuildProlongation();
+
+   bool store_matrices = false;
+
+   // Store the matrix L^-1 B  and Vector L^-1 l
+   // where G = L L^t
+   Array<ComplexDenseMatrix * > Bmat;
+   Array<Vector * > fvec;
+   Vector residuals;
+
+private:
+
+public:
+
+   /// Creates bilinear form associated with FE spaces @a *fespaces.
+   ComplexDPGWeakForm()
+   {
+      height = 0.;
+      width = 0;
+   }
+
+   ComplexDPGWeakForm(Array<FiniteElementSpace* > & fes_,
+                      Array<FiniteElementCollection *> & fecol_)
+   {
+      SetSpaces(fes_,fecol_);
+   }
+
+   void SetTestFECollVdim(int test_fec, int vdim)
+   {
+      test_fecols_vdims[test_fec] = vdim;
+   }
+
+   void SetSpaces(Array<FiniteElementSpace* > & fes_,
+                  Array<FiniteElementCollection *> & fecol_)
+   {
+      trial_fes = fes_;
+      test_fecols = fecol_;
+      test_fecols_vdims.SetSize(test_fecols.Size());
+      test_fecols_vdims = 1;
+      nblocks = trial_fes.Size();
+      mesh = trial_fes[0]->GetMesh();
+
+      IsTraceFes.SetSize(nblocks);
+      // Initialize with False
+      IsTraceFes = false;
+      for (int i = 0; i < nblocks; i++)
+      {
+         IsTraceFes[i] =
+            (dynamic_cast<const H1_Trace_FECollection*>(trial_fes[i]->FEColl()) ||
+             dynamic_cast<const ND_Trace_FECollection*>(trial_fes[i]->FEColl()) ||
+             dynamic_cast<const RT_Trace_FECollection*>(trial_fes[i]->FEColl()));
+      }
+      Init();
+   }
+
+   // Get the size of the bilinear form of the ComplexDPGWeakForm
+   int Size() const { return height; }
+
+   // Pre-allocate the internal SparseMatrix before assembly.
+   void AllocateMatrix() { if (mat_r == NULL) { AllocMat(); } }
+
+   ///  Finalizes the matrix initialization.
+   void Finalize(int skip_zeros = 1);
+
+   /// Returns a reference to the sparse matrix:  \f$ M \f$
+   BlockMatrix &BlockMat_r()
+   {
+      MFEM_VERIFY(mat_r, "mat_r is NULL and can't be dereferenced");
+      return *mat_r;
+   }
+   BlockMatrix &BlockMat_i()
+   {
+      MFEM_VERIFY(mat_i, "mat_i is NULL and can't be dereferenced");
+      return *mat_i;
+   }
+
+   /// Returns a reference to the sparse matrix of eliminated b.c.: \f$ M_e \f$
+   BlockMatrix &BlockMatElim_r()
+   {
+      MFEM_VERIFY(mat_e_r, "mat_e is NULL and can't be dereferenced");
+      return *mat_e_r;
+   }
+
+   BlockMatrix &BlockMatElim_i()
+   {
+      MFEM_VERIFY(mat_e_i, "mat_e is NULL and can't be dereferenced");
+      return *mat_e_i;
+   }
+
+   /// Adds new Trial Integrator. Assumes ownership of @a bfi.
```

Comment:
Comment should mention that m/n are indices into the blocks

Reply:
done!